File: core/src/main/java/ru/kuchanov/scpcore/monetization/model/Item.java
package ru.kuchanov.scpcore.monetization.model;

import com.google.gson.GsonBuilder;

/**
 * Created by mohax on 14.01.2017.
 * <p>
 * for scp_ru
 */
public class Item {

    public PurchaseData purchaseData;
    private final String signature;
    public String sku;
    private final String continuationToken;

    public Item(
            final String purchaseData,
            final String signature,
            final String sku,
            final String continuationToken
    ) {
        super();
        this.purchaseData = new GsonBuilder().create().fromJson(purchaseData, PurchaseData.class);
        this.signature = signature;
        this.sku = sku;
        this.continuationToken = continuationToken;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        final Item item = (Item) o;

        return sku.equals(item.sku);
    }

    @Override
    public int hashCode() {
        return sku.hashCode();
    }

    @Override
    public String toString() {
        return "Item{" +
                "purchaseData='" + purchaseData + '\'' +
                ", signature='" + signature + '\'' +
                ", sku='" + sku + '\'' +
                ", continuationToken='" + continuationToken + '\'' +
                '}';
    }
}

File: core/src/main/java/ru/kuchanov/scpcore/monetization/model/Item.kt
package ru.kuchanov.scpcore.monetization.model

import com.google.gson.GsonBuilder

/**
 * Created by mohax on 14.01.2017.
 */
data class Item(
        val rawPurchaseData: String = "{ \"purchaseToken\":\"\"}",
        val signature: String = "",
        var sku: String,
        val continuationToken: String = ""
) {

    var purchaseData: PurchaseData

    init {
        this.purchaseData = GSON.fromJson(rawPurchaseData, PurchaseData::class.java)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val item = other as Item?

        return sku == item!!.sku
    }

    override fun hashCode(): Int {
        return sku.hashCode()
    }

    companion object {
        private val GSON = GsonBuilder().create()
    }
}
